test/extensions/filters/http/jwt_authn/verifier_test.cc

```diff
@@ -0,0 +1,293 @@
+  int idx = 0;
+  NiceMock<MockAuthFactory> mock_factory_;
+  VerifyContextPtr context_;
+};
+
+// Deeply nested anys that ends in provider name
+TEST_F(VerifierTest, DeeplyNestedAnys) {
+  const char config[] = R"(
+providers:
+  example_provider:
+    issuer: https://example.com
+    audiences:
+    - example_service
+    - http://example_service1
+    - https://example_service2/
+    remote_jwks:
+      http_uri:
+        uri: https://pubkey_server/pubkey_path
+        cluster: pubkey_cluster
+    forward_payload_header: sec-istio-auth-userinfo
+    from_params:
+    - jwta
+    - jwtb
+    - jwtc
+rules:
+- match: { path: "/match" }
+  requires:
+    requires_any:
+      requirements:
+      - requires_any:
+          requirements:
+          - requires_any:
+              requirements:
+              - provider_name: "example_provider"
+)";
+  MessageUtil::loadFromYaml(config, proto_config_);
+  createVerifier();
+
+  auto mock_auth = std::make_unique<MockAuthenticator>();
+  EXPECT_CALL(*mock_auth.get(), doVerify(_, _, _))
+      .WillOnce(Invoke(
+          [](Http::HeaderMap&, std::vector<JwtLocationConstPtr>*,
+             std::function<void(const Status& status)>* callback) { (*callback)(Status::Ok); }));
+  EXPECT_CALL(*mock_auth.get(), sanitizePayloadHeaders(_)).Times(1);
+  mock_auths_.push_back(std::move(mock_auth));
+
+  EXPECT_CALL(mock_cb_, onComplete(_, _)).WillOnce(Invoke([](const Status& status, VerifyContext&) {
+    ASSERT_EQ(status, Status::Ok);
+  }));
+  auto headers = Http::TestHeaderMapImpl{
+      {":path", "/match?jwta=" + std::string(GoodToken) + "&jwtb=" + std::string(ExpiredToken)},
+  };
+  context_ = VerifyContext::create(headers, &mock_cb_);
+  verifier_->verify(*context_);
+}
+
+// require alls that just ends
+TEST_F(VerifierTest, CanHandleUnexpectedEnd) {
+  const char config[] = R"(
+providers:
+  example_provider:
+    issuer: https://example.com
+    audiences:
+    - example_service
+    - http://example_service1
+    - https://example_service2/
+    remote_jwks:
+      http_uri:
+        uri: https://pubkey_server/pubkey_path
+        cluster: pubkey_cluster
+    forward_payload_header: sec-istio-auth-userinfo
+rules:
+- match: { path: "/match" }
+  requires:
+    requires_all:
+      requirements:
+      - requires_all:
+)";
+  MessageUtil::loadFromYaml(config, proto_config_);
+  createVerifier();
+
+  auto mock_auth = std::make_unique<MockAuthenticator>();
+  EXPECT_CALL(*mock_auth.get(), doVerify(_, _, _)).Times(0);
+  mock_auths_.push_back(std::move(mock_auth));
+
+  EXPECT_CALL(mock_cb_, onComplete(_, _)).WillOnce(Invoke([](const Status& status, VerifyContext&) {
+    ASSERT_EQ(status, Status::Ok);
+  }));
+  auto headers = Http::TestHeaderMapImpl{{":path", "/match"}};
+  context_ = VerifyContext::create(headers, &mock_cb_);
+  verifier_->verify(*context_);
+}
+
+TEST_F(VerifierTest, TestRequiresAll) {
+  MessageUtil::loadFromYaml(RequiresAllConfig, proto_config_);
```

Comment:
We need more tests for "RequaresAll" with following
1) one of required token is missing
2) one of required token failed

Reply:
Additional tests added.